### PR TITLE
Add NET 5.0 testing

### PR DESCRIPTION
**Pull request by pakrym**

### CODE REVIEW

File: eng/pipelines/templates/jobs/archetype-sdk-client.yml
Comment (weshaggard) [inline eng/pipelines/templates/jobs/archetype-sdk-client.yml:35]: What does this doe for us?

**Comment by weshaggard** [comment]
Don't you want to also add or update some of the existing test targets https://github.com/Azure/azure-sdk-for-net/blob/master/eng/pipelines/templates/jobs/archetype-sdk-tests.yml#L134 to be net5?

File: eng/pipelines/templates/jobs/archetype-sdk-client.yml
Comment (pakrym) [reply eng/pipelines/templates/jobs/archetype-sdk-client.yml]: Installs 5.0, it's not on agents right now.

**Comment by pakrym** [comment]
>Don't you want to also add or update some of the existing test targets https://github.com/Azure/azure-sdk-for-net/blob/master/eng/pipelines/templates/jobs/archetype-sdk-tests.yml#L134 to be net5?

Yep on it.

File: eng/Directory.Build.Data.targets
Comment (pakrym) [inline eng/Directory.Build.Data.targets:158]: The package name is somehow case sensitive now.

File: eng/pipelines/templates/jobs/archetype-sdk-client.yml
Comment (pakrym) [inline eng/pipelines/templates/jobs/archetype-sdk-client.yml:152]: Let dotnet host reuse existing frameworks on the machine.

File: sdk/cognitiveservices/Vision.ContentModerator/tests/Helpers/Constants.cs
Comment (pakrym) [inline sdk/cognitiveservices/Vision.ContentModerator/tests/Helpers/Constants.cs:528]: .NET 5 enables more warnings

File: sdk/servicebus/Microsoft.Azure.ServiceBus/tests/Microsoft.Azure.ServiceBus.Tests.csproj
Comment (pakrym) [inline sdk/servicebus/Microsoft.Azure.ServiceBus/tests/Microsoft.Azure.ServiceBus.Tests.csproj:5]: THis project has Program.Main of it's own.

**Comment by pakrym** [comment]
Filed https://github.com/microsoft/azure-pipelines-tasks/issues/13895

**Review by kasobol-msft — APPROVED** [review]
Storage looks good.

File: sdk/cognitiveservices/Vision.ContentModerator/tests/Helpers/Utilities.cs
Comment (heaths) [inline sdk/cognitiveservices/Vision.ContentModerator/tests/Helpers/Utilities.cs:655]: It concerns me in how many places this antipattern exists. Perhaps we need a best practices section of our guidelines that warn against stuff like this with reasons.

File: sdk/core/Azure.Core.TestFramework/src/IgnoreOnNet5Attribute.cs
Comment (heaths) [inline sdk/core/Azure.Core.TestFramework/src/IgnoreOnNet5Attribute.cs:12]: Why? That signals incompatibility that needs to be addressed. Maybe require a link to a GitHub issue to address.

File: sdk/core/Azure.Core.TestFramework/src/IgnoreOnNet5Attribute.cs
Comment (heaths) [inline sdk/core/Azure.Core.TestFramework/src/IgnoreOnNet5Attribute.cs:14]: Should be inheritable like other test attributes. Both Core and Key Vault use inheritance a lot.

File: sdk/core/Azure.Core.TestFramework/src/IgnoreOnNet5Attribute.cs
Comment (heaths) [inline sdk/core/Azure.Core.TestFramework/src/IgnoreOnNet5Attribute.cs:21]: Assert this isn't null or empty, and maybe just require a link to actually fix the incompatibility.

File: sdk/core/Azure.Core.TestFramework/src/IgnoreOnNet5Attribute.cs
Comment (heaths) [inline sdk/core/Azure.Core.TestFramework/src/IgnoreOnNet5Attribute.cs:29]: Extra apostrophe after can't

File: sdk/core/Azure.Core/src/Shared/DiagnosticScope.cs
Comment (heaths) [inline sdk/core/Azure.Core/src/Shared/DiagnosticScope.cs:114]: Not: helpful for reviews if you keep the corresponding comment that VS would've added.

File: sdk/keyvault/Azure.Security.KeyVault.Keys/tests/AesCryptographyProviderTests.cs
Comment (heaths) [inline sdk/keyvault/Azure.Security.KeyVault.Keys/tests/AesCryptographyProviderTests.cs:141]: I see you did include a link. I'd just call "reason" "issueLink" instead to reinforce that.

File: sdk/keyvault/samples/keyvaultproxy/tests/AzureSamples.Security.KeyVault.Proxy.Tests.csproj
Comment (heaths) [inline sdk/keyvault/samples/keyvaultproxy/tests/AzureSamples.Security.KeyVault.Proxy.Tests.csproj:4]: This project is a sample meant to standalone on Samples Browser. That property won't exist normally. This change needs to be reverted.

File: sdk/search/Azure.Search.Documents/tests/FieldBuilderTests.cs
Comment (heaths) [inline sdk/search/Azure.Search.Documents/tests/FieldBuilderTests.cs:420]: Why did you change this? I used the same format that ArgumentExceptions do by design.

File: sdk/search/Azure.Search.Documents/tests/FieldBuilderTests.cs
Comment (heaths) [inline sdk/search/Azure.Search.Documents/tests/FieldBuilderTests.cs:445]: Same here.

File: sdk/search/Microsoft.Azure.Search/tests/Tests/FieldBuilderTests.cs
Comment (heaths) [inline sdk/search/Microsoft.Azure.Search/tests/Tests/FieldBuilderTests.cs:420]: I would like to know why all these were changed. I intentionally used the same format as ArgumentExceptions.

**Review by heaths — CHANGES_REQUESTED** [review]
Some suggestions, but I have open questions about some changes.

File: sdk/cognitiveservices/Vision.ContentModerator/tests/Helpers/Utilities.cs
Comment (RezaJooyandeh) [reply sdk/cognitiveservices/Vision.ContentModerator/tests/Helpers/Utilities.cs:655]: It is also weird to see many instances of a `try/catch` that catch and just rethrow.

File: sdk/communication/Shared/src/HMACAuthenticationPolicy.cs
Comment (drwill-ms) [inline sdk/communication/Shared/src/HMACAuthenticationPolicy.cs:48]: What does this do?

**Review by drwill-ms — COMMENTED** [review]
Thanks for tagging me. I don't have any concerns.

File: eng/Directory.Build.Data.props
Comment (heaths) [reply eng/Directory.Build.Data.props:89]: Nevermind on 2. I wasn't considering we test in parallel. For 1, I'm assuming because you added activity support, yes? But should we also test on 3.1 since it is LTS?

File: global.json
Comment (maririos) [inline global.json:6]: Will [this](https://github.com/Azure/azure-sdk-for-net/blob/master/CONTRIBUTING.md#prerequisites) need to be updated too?

File: sdk/eventhub/Azure.Messaging.EventHubs.Shared/tests/Core/TimeSpanExtensionsTests.cs
Comment (jsquire) [inline sdk/eventhub/Azure.Messaging.EventHubs.Shared/tests/Core/TimeSpanExtensionsTests.cs:92]: Good catch!   This definitely should have been in here.

**Review by jsquire — APPROVED** [review]
Event Hubs and Service Bus parts look good to me.

File: sdk/search/Azure.Search.Documents/tests/FieldBuilderTests.cs
Comment (pakrym) [reply sdk/search/Azure.Search.Documents/tests/FieldBuilderTests.cs:420]: ArgumentExceptions message format has changed in 3.1 so all assets that were using an exact format have to be changed.

File: sdk/communication/Shared/src/HMACAuthenticationPolicy.cs
Comment (pakrym) [reply sdk/communication/Shared/src/HMACAuthenticationPolicy.cs:48]: `alg.Hash` is marked as nullable while `Convert.ToBase64String` argument is marked as non-nullable. This tells C# compiler that we are positive that `alg.Hash` won't be null in this particular case.

File: global.json
Comment (pakrym) [reply global.json:6]: Great catch!

File: eng/Directory.Build.Data.props
Comment (pakrym) [reply eng/Directory.Build.Data.props:89]: We have 5.0 specific features (new activity, sync HTTP client) that we want to test. We had nothing like that in 3.1 and we are already testing one LTS. So now it's LTS + latest.

Comment (pakrym) [reply eng/Directory.Build.Data.props:89]: I'm reluctant to increase the matrix even more because of how much CI time it would take

File: eng/Directory.Build.Data.props
Comment (heaths) [inline eng/Directory.Build.Data.props:88]: Given some discussions on Twitter about this (with Immo et. al.) we might want to use "net5.0" here. "net5" works, but in another 5 years or so may pose problems.

File: sdk/keyvault/samples/keyvaultproxy/tests/AzureSamples.Security.KeyVault.Proxy.Tests.csproj
Comment (heaths) [reply sdk/keyvault/samples/keyvaultproxy/tests/AzureSamples.Security.KeyVault.Proxy.Tests.csproj:4]: An alternative fix would be to initialize `RequiredTargetFrameworks` here if not already initialized. These need to work standalone when downloaded from Samples Browser. I spent way too much time on these (would've been easier in a separate repo; cc @weshaggard) to make sure they work in this repo as well as separately.

**Review by heaths — APPROVED** [review]

**Comment by heaths** [comment]
Given questions answered, changing my review to unblock pending some changes.

File: eng/Directory.Build.Data.props
Comment (weshaggard) [reply eng/Directory.Build.Data.props]: I tend to agree that we should limit our test frameworks to what we feel is the biggest bang for the buck and not test all combinations.

File: global.json
Comment (weshaggard) [inline global.json:6]: Given that this just released are we sure we want to move the baseline to this version so quickly? I can understand starting to test this in CI but forcing everyone in the repo to move will cause some issues. 

@AlexGhiondea what are your thoughts here?

**Review by maririos — APPROVED** [review]
App Config, TA, and FR LGTM. 
Thanks for the change!!

**Comment by weshaggard** [comment]
@allenjzhang @markcowl @m-nash Do you guys foresee any issues with moving the baseline toolset up to .NET 5? Will mgmt library builds still function.

@pakrym @ShivangiReja I expect this to have an impact on the SDK automation pipeline work as we will now need make sure we install .NET 5 SDK as part of the init script.

File: eng/pipelines/templates/jobs/archetype-sdk-client.yml
Comment (weshaggard) [reply eng/pipelines/templates/jobs/archetype-sdk-client.yml]: Ok this will likely start to cause some reliability issues that we were seeing in the past, so we should keep an eye on this. Looks like they are working on .NET 5 support https://github.com/actions/virtual-environments/issues/1891.

File: global.json
Comment (pakrym) [reply global.json:6]: Considering we are adding net5.0 targets to all test projects 5.0 SDK is required to be able to test locally.

File: eng/pipelines/templates/jobs/archetype-sdk-tests.yml
Comment (weshaggard) [inline eng/pipelines/templates/jobs/archetype-sdk-tests.yml:132]: Is there a reason you picked windows?

File: global.json
Comment (weshaggard) [reply global.json:6]: Lets ensure we have agreement of moving the baseline up but if we have any blockers we should consider allowing rollforward on major with some detection that will add net5.0 as a test target if we are running on that framework.

File: eng/pipelines/templates/jobs/archetype-sdk-tests.yml
Comment (pakrym) [reply eng/pipelines/templates/jobs/archetype-sdk-tests.yml:132]: I don't have a good reason why any other platform would be better.

Comment (kasobol-msft) [reply eng/pipelines/templates/jobs/archetype-sdk-tests.yml:132]: Storage tests on Linux are significantly slower. That's +1 for windows (and MacOS).

File: global.json
Comment (pakrym) [reply global.json:6]: Definitely. I will hold this PR off as long as needed to get everyone on board.

File: eng/pipelines/templates/jobs/archetype-sdk-tests.yml
Comment (heaths) [reply eng/pipelines/templates/jobs/archetype-sdk-tests.yml:132]: linux and macOS are generally faster, though. That said, net5.0 is supposed to bring a lot of perf improvements to .NET in general, so they may still finish before net461 or even netcoreapp2.1.

**Review by RezaJooyandeh — APPROVED** [review]
Thanks for the change, communication part looks good to me 😊

File: eng/pipelines/templates/jobs/archetype-sdk-tests.yml
Comment (markcowl) [reply eng/pipelines/templates/jobs/archetype-sdk-tests.yml:132]: It would be nice, as a goal, to have tests on multiple platforms for our most used target framework

**Review by markcowl — APPROVED** [review]
LGTM - other than the merge conflicts.  In the future, I think .Net 5 on multiple platforms should be a goal

File: eng/pipelines/templates/jobs/archetype-sdk-client.yml
Comment (weshaggard) [inline eng/pipelines/templates/jobs/archetype-sdk-client.yml:98]: Do you really want to switch this to verify everything in the core pipeline?

File: eng/pipelines/templates/steps/install-dotnet.yml
Comment (weshaggard) [inline eng/pipelines/templates/steps/install-dotnet.yml:5]: I believe that DevOps has included .NET 5.0 on the box so lets please make sure that this step doesn't continue to download and install it again. We moved away from this step to avoid that networking reliability issues and I'd hate to have that come back.

File: eng/pipelines/templates/steps/install-dotnet.yml
Comment (weshaggard) [inline eng/pipelines/templates/steps/install-dotnet.yml:8]: Is this new? How was it being handled before?

**Review by weshaggard — APPROVED** [review]
I left a few comments, in particular the one about always causing an install to happen, that we should investigate a little before merging this PR but otherwise the changes look good.

File: eng/pipelines/templates/jobs/archetype-sdk-client.yml
Comment (pakrym) [reply eng/pipelines/templates/jobs/archetype-sdk-client.yml:98]: Yes, because only the core pipeline is started when the generator/snippet updater/targes are updated and we would like to make sure everything is up to date.

File: eng/pipelines/templates/steps/install-dotnet.yml
Comment (pakrym) [reply eng/pipelines/templates/steps/install-dotnet.yml:5]: Nope, I've tried and it's not available yet.

File: eng/pipelines/templates/steps/install-dotnet.yml
Comment (pakrym) [reply eng/pipelines/templates/steps/install-dotnet.yml:8]: Before we disabled the multi-level lookup and installed all versions of the runtime we need. This way we can only install what's not on the VM and let the dotnet.exe fallback to program files when required.

File: eng/pipelines/templates/steps/install-dotnet.yml
Comment (weshaggard) [reply eng/pipelines/templates/steps/install-dotnet.yml:5]: It is supposed to be coming soon, see https://github.com/actions/virtual-environments/issues/1891. We could consider waiting for that, but if we get this merged in before that work is ready we should file an issue to follow-up and remove this explicit install step once it ends up on the image.

File: eng/pipelines/templates/steps/install-dotnet.yml
Comment (pakrym) [reply eng/pipelines/templates/steps/install-dotnet.yml:8]: https://github.com/Azure/azure-sdk-for-net/issues/17118

File: eng/pipelines/templates/steps/install-dotnet.yml
Comment (pakrym) [reply eng/pipelines/templates/steps/install-dotnet.yml:5]: #17118

**Review by chamons — APPROVED** [review]

**Review by schaabs — APPROVED** [review]

**Review by christothes — APPROVED** [review]

**Pull request merged**
